### DISCUSSION TITLE
Reuse source after `exit 0` in `rotate.merge`

### DISCUSSION THREAD

**Discussion by noctuum** [savonet/liquidsoap#2860] [User Support]
Hello.

Please help to find out how to reuse source in `rotate.merge` after it's exit. As it states in docs, [rotate.merge](https://www.liquidsoap.info/doc-2.1.3/reference.html#rotate.merge) should loops a sequence of sources. But after source `exit 0` on the end of track, stream goes `safe_blank` instead of looping sources.

Some info:
```
$ uname -a
Linux localhost 6.1.0-1-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.4-1 (2023-01-07) x86_64 GNU/Linux

$ liquidsoap --version
Liquidsoap 2.1.3
```

LS code:
```
#!/usr/bin/liquidsoap
log.file.path.set("/home/user/radio/radio.log")
log.stdout.set(true)

def input.airsonic(~id="input.airsonic",
		~restart=false,~restart_on_error=false,
		~buffer=0.2,~max=20.,s) =
	input.external.rawaudio(id=id,restart=restart,
		restart_on_error=restart_on_error,
		buffer=buffer,max=max,
		"/home/user/radio/streamer #{string.quote(s)}")
end

trackids = (process.read.lines("cat /home/user/radio/total_songs.list"))

def input.airsonic.random(~id="input.airsonic.random",
		~restart=false,~restart_on_error=false) =
	input.airsonic(id=id,restart=restart,
		restart_on_error=restart_on_error,
		list.pick(list.shuffle(trackids)))
end

radio = rotate.merge([input.airsonic.random(),input.airsonic.random()])
radio = mksafe(radio)

output.icecast(%fdkaac(channels=2, samplerate=44100,
		bandwidth="auto", bitrate=128, afterburner=true,
		aot="mpeg4_he_aac_v2", transmux="adts", sbr_mode=true),
	host = "x",
	port = x,
	password = "x",
	mount = "x",
radio)
```

With current code, LS plays two random tracks and goes `safe_blank` after each of sources exit. If i try to set
`rotate.merge([input.airsonic.random(restart=true),input.airsonic.random(restart=true)])`
only first source plays and it constantly repeats 1 track, so it does not perform `list.pick()` every time as expected.

Is there a way to reuse source without restart argument? Or to re-read variable on source repeat? Or maybe there is another approach?

**Comment by toots** [comment]
Thanks for reporting. The problem is that `liquidsoap` does not insert a track mark when the process restarts so it appears as a single track. I just changed this behavior in `rolling-release-v2.1.x`, which will be released as `2.1.4`. With this release, you should start seeing process restart behave as a new track, which should trigger the behavior you are expecting.

Once the build here: https://github.com/savonet/liquidsoap/actions/runs/4065980109 finishes, the release should be available for testing here: https://github.com/savonet/liquidsoap/releases/tag/rolling-release-v2.1.x. `2.1.4` is expected to be released soon.

**Comment by noctuum** [comment]
Great! Thank you! It now switches between sources, so I can even apply crossfade.
But it turned out that now I have two sources with static ID obtained from `#{string.quote(s)}`. So only two tracks are playing on repeat. Could you please prompt me how can I get new `#{string.quote(s)}` on every source change?

**Comment by toots** [comment]
Yeah you need a string getter for this so that the process string is reload each time it restarts. I think this should be fixed in the latest rolling release and `main`. You should try this with it now:

```
log.file.path.set("/home/user/radio/radio.log")
log.stdout.set(true)

def input.airsonic(~id="input.airsonic",
		~restart=false,~restart_on_error=false,
		~buffer=0.2,~max=20.,s) =
	input.external.rawaudio(id=id,restart=restart,
		restart_on_error=restart_on_error,
		buffer=buffer,max=max,
		{"/home/user/radio/streamer #{string.quote(s())}"})
end

trackids = (process.read.lines("cat /home/user/radio/total_songs.list"))

def input.airsonic.random(~id="input.airsonic.random",
		~restart=false,~restart_on_error=false) =
	
	input.airsonic(id=id,restart=restart,
		restart_on_error=restart_on_error,
		{list.pick(list.shuffle(trackids))})
end

radio = rotate.merge([input.airsonic.random(),input.airsonic.random()])
radio = mksafe(radio)

output.icecast(%fdkaac(channels=2, samplerate=44100,
		bandwidth="auto", bitrate=128, afterburner=true,
		aot="mpeg4_he_aac_v2", transmux="adts", sbr_mode=true),
	host = "x",
	port = x,
	password = "x",
	mount = "x",
radio)
```

**Comment by noctuum** [comment]
Thank you for quick response and for helping!

I have update the package and change code to the one you have provided and now two different tracks play, then stream goes `safe_blank`.

Meanwhile I was requested to change the input from raw audio source to http source (flac/mp3) so LS will receive metadata automatically.
In this case I guess for getting new `#{string.quote(s)` everytime, I can use `request.dynamic` for generating a new http link containing new random ID.
But this is surely out of the scope, since header of the topic was actually fixed. I may open a new thread later for other question. Thank you very much for your help!
